### DISCUSSION TITLE
Anticipate any benefit from using esbuild somehow?

### DISCUSSION THREAD

**Discussion by martaver** [vercel/next.js#16152] [Help]
A question directed to the experts (the nextjs team)...

Ever since I saw [esbuild](https://github.com/evanw/esbuild), I'm having weird dreams where webpack adopts it in some way and becomes 100x faster.

Do you anticipate being able to realistically leverage its capabilities in any way with webpack, or otherwise, in order to achieve a significant DX performance improvement? Why or why not?

**Comment by leerob** [comment]
There a lot of exciting innovation in this space. Esbuild, Skypack, Vite, Roam, Parcel 2, etc. My guess is the focus in Next will be on Webpack 5 adoption, and then as some of these tools mature, explore adding them into the platform.

**Comment by eezing** [comment]
esbuild is incredibly fast. Here’s an interesting project: https://github.com/privatenumber/esbuild-loader

**Comment by kirill-konshin** [comment]
Recently integrated it in one of the projects with 6K modules. Can't confirm 100x increase in performance, but 10x for sure. Sometimes more depending on phase of the moon probably ;)

**Comment by alan-cooney** [comment]
Likewise seeing c. 80x improvement in build times for another project - it would be great to add this in

**Comment by nandorojo** [comment]
My Next.js app in a monorepo takes 15 minutes to build. I would do just about anything to see something as fast at esbuild used for builds.

**Comment by mmkal** [comment]
Here's an issue on esbuild-loader where some folks have tried swapping next-babel-loader for esbuild-loader: https://github.com/privatenumber/esbuild-loader/issues/84. It looks like it kind-of works, with some edge cases unhandled that prevent it from being ready for primetime - might be that some help from next.js maintainers is needed to get it working properly. This would be a _huge_ win. CC @timneutkens

**Reply by 205g0** [reply]
What are those edge cases? _Edit: just skimmed that link and yeah, there're the edge cases..._

**Comment by nandorojo** [comment]
Here's an elaborated example of Next.js + ESBuild: https://github.com/privatenumber/esbuild-loader-examples/tree/master/examples/next

Anyone who can test this with me would be great. Would love to identify edge cases.

**Reply by s-kris** [reply]
https://github.com/privatenumber/esbuild-loader/issues/114
I'll update you on this

**Reply by s-kris** [reply]
I got it build successfully by removing next css plugin but I don't crazy fast build times. Just 2x.
Normal next build: 150s
Esbuild-loader: 70s

**Reply by kirill-konshin** [reply]
2x is A LOT.

**Reply by s-kris** [reply]
Yes, absolutely. But when looking at esbuild's scale of 100x, there's a lot of scope.

**Comment by jamesarosen** [comment]
It's not just a DX improvement. [ESBuild](https://github.com/evanw/esbuild/commit/d5f7fdd8f71ad398dae591d5270aac71bac17590) (and [Rollup](https://rollupjs.org/guide/en/#tree-shaking), but not Webpack) can do advanced tree-shaking, which reduces resulting package sizes. For example, given
```js
// node_modules/some-library/index.js
export { default as foo } from './src/foo';
export { default as bar } from './src/bar';
export { default as baz } from './src/baz';

// components/my-component.jsx
import { foo, bar } from 'some-library'
```
ESBuild and Rollup can exclude `node_modules/some-library/src/baz.js` from the build, but Webpack includes it.

**Reply by kirill-konshin** [reply]
@jamesarosen actually Webpack also can do tree shaking, but maybe not as advanced as others.

**Reply by ianldgs** [reply]
I just tested webpack tree shaking the other day... and it worked! Wasn't specifically this situation, but try to set the `mode` to `production` and check if it will include the code.
